File: Middle/Array and String/15.py
class Solution:
    def binary(self, nums, val, left, right):
        """
        返回第一个不小于 val 的值的序号
        """
        if val > nums[right] :
            return -1
        while left < right:
            mid = (left+right)//2
            if nums[mid] >= val:
                right = mid
            else:
                left = mid+1
        return right

    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums)<3:
            return []


        nums.sort()
        result = []
        nneg_start = self.binary(nums, 0, 0, len(nums)-1)
        if nneg_start+2 < len(nums) and nums[nneg_start]==nums[nneg_start+2]==0:
            result.append([0, 0, 0])

        for i in range(nneg_start):
            if i==0  or nums[i]!=nums[i-1]:
                n_i = -nums[i]
                if nums[-1] >= 2*n_i:
                    k_right = self.binary(nums, 2*n_i, nneg_start, len(nums)-1) # =
                    k_left = self.binary(nums, n_i//2, nneg_start, k_right)     # !=
                    j_right = k_left                                            # =
                    j_left = i+1                                                # =
                elif nums[-1] >= n_i:
                    k_right = len(nums)-1
                    k_left = self.binary(nums, n_i//2, nneg_start, k_right)
                    j_right = k_left
                    j_left = self.binary(nums, n_i-nums[-1], i+1, nneg_start)
                elif nums[-1] >= n_i//2:
                    k_right = len(nums)-1
                    k_left = self.binary(nums, n_i//2, nneg_start, k_right)
                    j_right = k_left
                    j_left = self.binary(nums, n_i-nums[-1], nneg_start, j_right)
                else:
                    continue
                if j_left == k_right:
                    continue
                if nums[k_left] == n_i/2 and k_left!=k_right:
                    k_left += 1
                for j in range(j_left, j_right+1):
                    if j==j_left or nums[j]!=nums[j-1]:
                        k = self.binary(nums, -(nums[i]+nums[j]), k_left, k_right)
                        if k!=-1 and nums[i]+nums[j]+nums[k]==0:
                            result.append([nums[i], nums[j], nums[k]])

        return result

File: Middle/Array and String/15.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
private:
    int binary(vector<int>& nums, int val, int left, int right){
        if (val > nums[right])
            return -1;
        int mid;
        while (left < right){
            mid = int(left+right)/2;
            if (nums[mid]>=val)
                right = mid;
            else
                left = mid + 1;
        }
        return right;
    }

public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        if(nums.size() < 3)
            return result;
        int len = nums.size();
        sort(nums.begin(), nums.end());
        vector<int> item;
        int nneg_start = this->binary(nums, 0, 0, len-1);
        if (nneg_start+2 < len && nums[nneg_start]==nums[nneg_start+2] && nums[nneg_start]==0){
            item.clear();
            item.push_back(0);
            item.push_back(0);
            item.push_back(0);
            result.push_back(item);
        }

        for(int i=0;i<nneg_start;i++){
            if (!i || nums[i]!=nums[i-1]){
                int n_i = -nums[i];
                int k_left, k_right, j_left, j_right;
                if(nums[len-1] >= 2*n_i){
                    k_right = this->binary(nums, 2*n_i, nneg_start, len-1);
                    k_left = this->binary(nums, int(n_i/2), nneg_start, k_right);
                    j_right = k_left;
                    j_left = i+1;
                }
                else if(nums[len-1] >= n_i){
                    k_right = len-1;
                    k_left = this->binary(nums, int(n_i/2), nneg_start, k_right);
                    j_right = k_left;
                    j_left = this->binary(nums, n_i-nums[len-1], i+1, nneg_start);
                }
                else if(nums[len-1] >= int(n_i/2)){
                    k_right = len-1;
                    k_left = this->binary(nums, int(n_i/2), nneg_start, k_right);
                    j_right = k_left;
                    j_left = this->binary(nums, n_i-nums[len-1], nneg_start, j_right);
                }
                else
                    continue;
                if(j_left == k_right)
                    continue;
                if(nums[k_left]==int(n_i/2) && k_left!=k_right)
                    k_left += 1;

                for(int j=j_left;j<j_right+1;j++){
                    if(j==j_left || nums[j]!=nums[j-1]){
                        int k = this->binary(nums, -(nums[i]+nums[j]), k_left, k_right);
                        if(k!=-1 && nums[i]+nums[j]+nums[k]==0){
                            item.clear();
                            item.push_back(nums[i]);
                            item.push_back(nums[j]);
                            item.push_back(nums[k]);
                            result.push_back(item);
                        }
                    }
                }
            }
        }
        return result;
    }
};
